### PR TITLE
[react-redux] Improve Selector type to handle function with no OwnProps

### PR DESCRIPTION
**Pull request by VincentLanglet**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Cf issue : https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25804

### CODE REVIEW

**Comment by VincentLanglet** [comment]
I'm not sure it's the best way to do it.

The goal is to have a bivalent type : 
```
const foo: Selector<State, string> = (state: State) => ...
const bar: Selector<State, string, OwnProps> = (state: State, ownProps: OwnProps) => ...
```
With the errors : 
```
foo(state) // ok
foo(state, ownProps) // error expected 1 argument
bar(state) // error expected 2 argument
bar(state, ownProps) // ok
```

`(state: State, ownProps?: OwnProps)` is not doing the job

**Review by mDibyo — CHANGES_REQUESTED** [review]
@VincentLanglet I don't think this work as expected. For instance, the following does not raise any errors with either the existing definition or the new definition you added:
```ts
const f: Selector<State, Props, OwnProps> = (state: State) => {
    ...
}
```
It looks like Typescript thinks `(state: State): TProps` satisfies `(state: State, ownProps: TOwnProps) => TProps`, so seems impossible to enforce the bivalent type as you wanted to.

**Comment by VincentLanglet** [comment]
@alloy Why should I do new commit if I don't want to see this PR closed ? The only reviewer post a comment and I don't have answer from him. I think sometimes is wrong or not clear with this bot behaviour.

**Comment by VincentLanglet** [comment]
@DanielRosenwasser @rbuckton I have no news from @mDibyo 
I have tested this change in my project, and it's working well. And I have written tests.
What can I do more ?
Thanks

File: types/react-redux/index.d.ts [hunk 1]
```diff
@@ -352,7 +352,9 @@ export function connectAdvanced<S, TProps, TOwnProps, TFactoryOptions = {}>(
 export type SelectorFactory<S, TProps, TOwnProps, TFactoryOptions> =
     (dispatch: Dispatch<Action>, factoryOptions: TFactoryOptions) => Selector<S, TProps, TOwnProps>;
 
-export type Selector<S, TProps, TOwnProps> = (state: S, ownProps: TOwnProps) => TProps;
+export type Selector<S, TProps, TOwnProps = null> = TOwnProps extends null
```
Comment (uniqueiniquity) [inline types/react-redux/index.d.ts]: You'll actually want `extends null | undefined` to totally match your example.

File: types/react-redux/react-redux-tests.tsx [hunk 2]
```diff
@@ -1192,3 +1192,16 @@ function TestProviderContext() {
...
+function TestSelector() {
+    interface OwnProps { key?: string; }
+    interface State { key: string; }
+
+    const simpleSelect: Selector<State, string> = (state: State) => state.key;
+    const notSimpleSelect: Selector<State, string, OwnProps> = (state: State, ownProps: OwnProps) => ownProps.key || state.key;
+
+    const ownProps = {};
+    const state = { key: 'value' };
+    simpleSelect(state);
```
Comment (uniqueiniquity) [inline types/react-redux/react-redux-tests.tsx]: You can also add the following to test:
```ts
simpleSelect(state, ownProps); // $ExpectError
notSimpleSelect(state); // $ExpectError
```

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
@uniqueiniquity Thanks for your review. I made the changes.

**Review by uniqueiniquity — APPROVED** [review]

**Comment by mDibyo** [comment]
@VincentLanglet I am sorry for not getting back to you on this PR. :( I wanted to play with the example you posted, but forgot about it.

**Comment by VincentLanglet** [comment]
@mDibyo No problem !

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Pull request merged**